Allocate global variables before populating constants

glguy opened the pull request:
Constants can depend on the addresses of global variables

```c
int flag;
int * const flag_ptr = &flag;
```

glguy committed:
Add ability to allocate all globals

This is necessary when global constants depend on the addresses of global variables.

RyanGlScott commented on saw-script/src/SAWScript/Interpreter.hs:
```diff
@@ -1885,6 +1886,16 @@ disable_lax_loads_and_stores = do
+disable_alloc_all_globals = do
+  rw <- getTopLevelRW
+  putTopLevelRW rw { rwAllocAllGlobals = False }
+
```
I actually do think there is a use case for turning the flag off again. This feature makes SAW's LLVM backend slightly less sound with respect to mutable globals, in the sense that you can verify a function that writes to a mutable global without explicitly allocating it with `llvm_alloc_global`. I think a reasonable use case would be to do something like this:

```
enable_alloc_all_globals;
m <- llvm_load_module "...";
disable_alloc_all_globals;
```

This would allow loading the constants in the module that depend on mutable globals to succeed, but it would then ensure that subsequent calls to `llvm_verify` would need to use `llvm_alloc_global` as required.

sauclovian-g replied:
I guess that's true.

In that case I think we're better off having one function that takes a boolean...

(and if we had optional arguments, it should just be an optional argument to llvm_load_module, but we don't and aren't getting them anytime soon)

sauclovian-g replied:
It should, but do we want more of them? It doesn't matter that much but it's just adding more ugliness...

RyanGlScott commented on saw-script/src/SAWScript/Interpreter.hs:
```diff
@@ -2693,6 +2704,18 @@ primitives = Map.fromList
+    (pureVal enable_alloc_all_globals)
+    Experimental
+    [ "Enable allocation of all globals automatically. This is necessary when"
+    , " constants depend on the addresses of globals."
```
This shouldn't affect the behavior of the JVM or MIR backends at all. Besides the fact that this PR doesn't touch any code in those two backends, it is also worth nothing that both backends handle globals differently enough from the LLVM backend that it's not really meaningful to talk about how `enable_alloc_all_globals` would interact with them.

sauclovian-g replied:
I apparently thought it was updating shared code. N/M

RyanGlScott reviewed (commented):
Please add a test case and mention the new commands in the changelog.

kquick committed:
Add three global allocation options for LLVM.

Re-implemented from original glguy to accomodate other changes due to rebasing on
latest master.

kquick committed:
Fixup from previous rebasing.

kquick committed:
Add test2623 for new llvm_alloc_X_globals commands.

kquick committed:
Update SAW User Manual and CHANGELOG for new llvm_alloc_X_globals.

kquick committed:
Update import for changed SAWServer default global allocation stance.

RyanGlScott commented on doc/saw-user-manual/specification-based-verification.md:
```diff
@@ -1655,8 +1655,10 @@ leading to unsoundness in the presence of compositional verification. (For more
+Immutable (i.e. `const`) global variables are allocated implicitly by default and
+do not require a call to `llvm_alloc_global`, although this can be changed (see
+the [LLVM Globals and Static
+Initialization](#llvm-globals-and-static-initialization) below)..
```
```suggestion
Initialization](#llvm-globals-and-static-initialization) below).
```

RyanGlScott commented on doc/saw-user-manual/specification-based-verification.md:
```diff
@@ -1747,6 +1749,48 @@ point of a call to `f`. This specification also constrains `y` to prevent
+subsequent `llvm_verify`, SAW must first process the `_start` initialization code
+emitted by the compiler.  Here `flag_ptr` is initialized in `_start` which will
+attempt to write the address of `flag`, but since `flag` is mutable, it will not
+have a storage allocation yet.
```
While it's true that SAW simulates initialization of global variables, I think it's misleading to say that it's processing the `_start` function specifically. SAW will initialize globals regardless of whether a bitcode file contains a `_start` function or not, and indeed, many examples do not contain one. I'd favor just removing the mentions of `_start` here (and below in the `llvm_allocate_{all,no}_globals` bullet points).

kquick replied:
Fair enough.  the SAW error refers to `_start` though:

```
AssertionFailure :1:0: error: in _start
Global symbol not allocated
Details:
  Global symbol "flag" has no associated allocation
```
so I've updated the initialization to setup the What4 ExprBuilder explicitly instead of letting it default so this now reports:
```
AssertionFailure intTests/test2623/test.saw:75:8: error: in _SAW_setupLLVMCrucibleContext
Global symbol not allocated
Details:
  Global symbol "flag" has no associated allocation
```
which is better because (a) it isn't misleading by reporting `_start`, and (b) it provides the actual location in the SAW script which is a lot more helpful.

Also interestingly, the above `_start` reference only occurred for the `llvm_alloc_constant_globals` case, but `llvm_alloc_no_globals` reported `AserrtionFailure: internal: error: in _SAW_LLVM_verifyPrestate` which differed in both the location (`internal`) and the `in` reference.  In this case, the `in` reference provides better information, but the `internal` loc does not, so I've also updated that such that I now get `AssertionFailure intTests/test2623/test.saw:50:3: error: in _SAW_LLVM_verifyPrestate`, and now the two error reporting cases are better aligned information-wise.

RyanGlScott replied:
Ah, I forgot that SAW uses the name "`_start`" as a general source location for error messages. (I interpreted `_start` as the C function that compilers often auto-generate.) In that case, I guess we need to pick _some_ name for this location, and `_SAW_setupLLVMCrucibleContext` seems as fine of a choice as any.

RyanGlScott commented on doc/saw-user-manual/specification-based-verification.md:
```diff
@@ -1747,6 +1749,48 @@ point of a call to `f`. This specification also constrains `y` to prevent
+* `llvm_allocate_no_globals` specifies that no implicit allocation should be
+  performed and that all allocation will be handled by the `LLVMSetup`
+  explicitly.  This will not be possible if `_start` attempts to refer to the
+  allocations of any global variables.
```
Naming nit pick: I'm bothered by the fact that the first command uses the prefix `llvm_alloc`, while the latter two use the prefix `llvm_allocate`. I'd prefer that all commands be consistently named.

RyanGlScott commented on doc/saw-user-manual/specification-based-verification.md:
```diff
@@ -1747,6 +1749,48 @@ point of a call to `f`. This specification also constrains `y` to prevent
+* `llvm_allocate_no_globals` specifies that no implicit allocation should be
+  performed and that all allocation will be handled by the `LLVMSetup`
+  explicitly.  This will not be possible if `_start` attempts to refer to the
+  allocations of any global variables.
```
I really think we should emphasize that these two options are unsound in the general case, and they should be used sparingly. It might also help to list specific examples of programs that SAW will unsoundly verify if you use these options recklessly.

kquick replied:
Is `llvm_alloc_no_globals` itself unsound?  I think it's likely to lead simply to failures due to uninitialized memory which will require explicit allocation and initialization as described elsewhere, which has appropriate soundness and compositional discussions.  I can add an unsoundness warning here if you think it does belong, but I'm not sure this is the part that could be unsound.

RyanGlScott replied:
> Is `llvm_alloc_no_globals` itself unsound?

I haven't checked, but I fear that if you don't allocate immutable globals explicitly, then you can write a spec where a user allocations an immutable global but then has it point to a different value than what `llvm_global_initializer` would give, leading to the function returning something different than what it should.

kquick replied:
I don't disagree with the concern, but I was mostly asserting that allocation is due to other statements, not the `llvm_alloc_no_globals` itself.

I did make changes here to provide warnings and steer users in the direction of `llvm_global_initializer` without indicating it was required (as you mentioned elsewhere).  I also feel like our docs around these are pretty good (I haven't looked at our new doc branch yet).  The `llvm_alloc_no_globals` was added per your request; let me know if the current description is insufficient and if so if you have some verbiage you'd suggest for me?

RyanGlScott commented on saw-script/src/SAWScript/Interpreter.hs:
```diff
@@ -5410,6 +5435,21 @@ primitives = Map.fromList $
+  , prim "llvm_alloc_no_globals" "TopLevel ()"
+    (pureVal (llvm_set_global_alloc_mode LLVMAllocNoGlobals))
+    Experimental
+    [ "Disable automatic allocation of all globals." ]
```
Similarly, we should emphasize `llvm_alloc_{all,no}_globals`'s unsoundness here and include references to the relevant section of the SAW manual for readers to learn more.

RyanGlScott commented on intTests/test2623/test.saw:
```diff
@@ -0,0 +1,84 @@
+// immutable global pointer is handled by the `_start` portion of the program,
+// which SAW evaluates *before* running the LLVMSetup, so the aforementioned
+// operations are not yet effective, and thus the `_start` attempt to get the
+// allocation address of the mutable global variable fails.
```
Again, there's no `_start` function in either `foo.bc` or `bar.bc`, so I'd favor rewording this to just talk about how SAW initializes globals during the start of simulation.

RyanGlScott commented on intTests/test2623/test.saw:
```diff
@@ -0,0 +1,84 @@
+enable_experimental;
+
+
+// The "foo.bc" and "bar.bc" sources are identical except foo initializes two
```
Should this say "The `foo.c` and `bar.c` sources" instead? `foo.bc` and `bar.bc` aren't source code.

RyanGlScott commented on intTests/test2623/test.saw:
```diff
@@ -0,0 +1,84 @@
+// By default, SAW will implicitly allocate storage for immutable global
+// variables, but due to compositionality concerns, it requires explicit storage
+// allocation and initialization for mutable variables (via llvm_alloc_global and
+// corresponding llvm_points_to of llvm_global to an llvm_global_initializer)
```
Strictly speaking, you don't need to have a mutable global point to an `llvm_global_initializer`---you can just as well have it point to some other value.

RyanGlScott commented on intTests/test2623/test.saw:
```diff
@@ -0,0 +1,84 @@
+// llvm_alloc_all_globals, which ensures that Crucible allocates memory for *all*
+// globals (thereby allowing any _start initializations to succeed).
+// Alternatively, llvm_alloc_no_globals can be used to suppress all implicit
+// allocation as a way to disable any preconcieved assumptions of the target
```
```suggestion
// allocation as a way to disable any preconceived assumptions of the target
```

RyanGlScott commented on saw-central/src/SAWCentral/Crucible/LLVM/Builtins.hs:
```diff
@@ -1031,12 +1031,7 @@ setupGlobalAllocs cc mspec mem0 = foldM go mem0 $ mspec ^. MS.csGlobalAllocs
           gimap = view Crucible.globalInitMap mtrans
       case Map.lookup symbol gimap of
         Just (g, Right (mt, _)) -> ccWithBackend cc $ \bak ->
-          do when (L.gaConstant $ L.globalAttrs g) . throwMethodSpec mspec $ mconcat
-               [ "Global variable \""
-               , name
-               , "\" is not mutable"
-               ]
```
Why was this check removed? Surely we still want to check this in at least the `llvm_alloc_constant_globals` case?

kquick committed:
Initialize What4 ExprBuilder location ASAP.

kquick committed:
Better position reporting for verification setup.

kquick committed:
Update documentation and interpreter help with soundness cautions.

kquick committed:
Restore generation of errors for already allocated const globals.

kquick committed:
Update intTest for new error location reporting.

sauclovian-g commented:
Would you be willing to rearrange the manual changes against #3231? I rewrote the whole section they're in :-|

sauclovian-g commented:
(and in particular, some of the discussion of globals and global initializers in the old manual text was chasing off in the wrong direction, if not actually wrong; the new text has a completely different emphasis)

kquick commented:
@sauclovian-g , I'm willing to send you a patch for that branch.  Do we just let this land as documented until that PR overrides it, or will that create more conflict noise you'd rather avoid somehow?  I could also remove the doc updates from this PR if that's preferred.

sauclovian-g commented:
As long as we have corresponding updated text in the branch, keeping the version here will generate a conflict but one that's easy to deal with (just drop the changes)... actually trying to merge it is what I'd like to avoid.

So if you could send a patch that'd be great... it's probably not a good idea to push directly to my branch yet, but sending it out of band is fine, or if you'd prefer to push, make a branch off my branch and tell me where it is (no need to go to the trouble of filing a PR to my branch) and I'll make sure it gets in and doesn't get lost by accident.

RyanGlScott commented on doc/saw-user-manual/specification-based-verification.md:
```diff
@@ -1747,6 +1749,50 @@ point of a call to `f`. This specification also constrains `y` to prevent
+* `llvm_alloc_all_globals` specifies that *all* globals are to be implicitly
+  allocated, regardless of their mutability.  Using this will ensure that the
+  `_start` pre-processing can complete successfully, but this can yield unsound
+  results if the `LLVMSetup` code does not initialized values for proper
```
```suggestion
  results if the `LLVMSetup` code does not initialize values for proper
```

RyanGlScott commented on doc/saw-user-manual/specification-based-verification.md:
```diff
@@ -1747,6 +1749,50 @@ point of a call to `f`. This specification also constrains `y` to prevent
+
+* `llvm_alloc_no_globals` specifies that no implicit allocation should be
+  performed and that all allocation will be handled by the `LLVMSetup`
+  explicitly.  This will not be possible if `_start` attempts to refer to the
```
Some lingering references to `_start` here

RyanGlScott commented on intTests/test2623/test.saw:
```diff
@@ -0,0 +1,84 @@
+// Alternatively, llvm_alloc_no_globals can be used to suppress all implicit
+// allocation as a way to disable any preconceived assumptions of the target
+// program (but any _start references will then be problematic).
+
```
Some lingering references to `_start` here

kquick committed:
More rewording of initialization descriptions.
